torchdata/datapipes/iter/util/randomsplitter.py

```diff
@@ -0,0 +1,135 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import random
+from typing import Dict, Optional, TypeVar, Union
+
+from torchdata.datapipes import functional_datapipe
+from torchdata.datapipes.iter import IterDataPipe
+
+T = TypeVar("T")
+
+
+@functional_datapipe("random_split")
+class RandomSplitterIterDataPipe(IterDataPipe):
+    r"""
+    Randomly split samples from a source DataPipe into groups(functional name: ``random_split``).
+    Since there is no buffer, only ONE group of samples (i.e. one child DataPipe) can be iterated through
+    at any time. Attempts to iterate through multiple of them simultaneously will fail.
+
+    Note that by default, multiple iterations of this DataPipe will yield the same split for consistency across epochs.
+    You can invoke ``set_seed`` on the output(s) to update the seed whenever needed (such as per epoch).
+
+    Args:
+        source_datapipe: Iterable DataPipe being split
+        total_length: Length of the ``source_datapipe``
+        weights: Dict of weights; the length of this list determines how many output DataPipes there will be.
+        seed: random _seed used to determine the randomness of the split
+        target: Optional key (that must exist in ``weights``) to indicate the specific group to return.
+            If set to the default ``None``, returns ``List[IterDataPipe]``.
+            If target is specified, returns ``IterDataPipe``.
+
+    Example:
+        >>> from torchdata.datapipes.iter import IterableWrapper
+        >>> dp = IterableWrapper(range(10))
+        >>> train, valid = dp.random_split(total_length=10, weights={"train": 0.5, "valid": 0.5}, seed=0)
+        >>> list(train)
+        [2, 3, 5, 7, 8]
+        >>> list(valid)
+        [0, 1, 4, 6, 9]
+        >>> # You can also specify a target key if you only need a specific group of samples
+        >>> train = dp.random_split(total_length=10, weights={"train": 0.5, "valid": 0.5}, seed=0, target='train')
+        >>> list(train)
+        [2, 3, 5, 7, 8]
+        >>> # Be careful to use the same seed as before when specifying `target` to get the correct split.
+        >>> valid = dp.random_split(total_length=10, weights={"train": 0.5, "valid": 0.5}, seed=0, target='valid')
+        >>> list(valid)
+        [0, 1, 4, 6, 9]
+    """
+
+    def __new__(
+        cls,
+        source_datapipe: IterDataPipe,
+        weights: Dict[T, Union[int, float]],
+        seed,
+        total_length: Optional[int] = None,
+        target: Optional[T] = None,
+    ):
+        if total_length is None:
```

Comment:
This might be an issue for DataPipes which have only runtime length. Can we add TODO line to review this later on.

Reply:
Yea, this is partially why made `total_length` mandatory, so that users have to make a conscious decision to call `len` or provide an integer if it is known.

I will add a note in the docstring to encourage users to provide one as well as add a TODO line.